# Interim solution to address expenditure data issue for March 15 presentation

* Using data from Data Link.

In [1]:
import numpy as np
import pandas as pd
import TIRCP_functions
from siuba import *
from calitp import *

GCS_FILE_PATH = "gs://calitp-analytics-data/data-analyses/tircp/"

pd.options.display.max_columns = 50
pd.options.display.max_rows = 120
pd.set_option('display.max_colwidth', None)

/opt/conda/lib/python3.9/site-packages/geopandas/_compat.py:111: UserWarning: The Shapely GEOS version (3.10.2-CAPI-1.16.0) is incompatible with the GEOS version PyGEOS was compiled with (3.10.1-CAPI-1.16.0). Conversions between both will be slow.
E0310 21:25:22.269069078    1284 fork_posix.cc:70]           Fork support is only compatible with the epoll1 and poll polling strategies
E0310 21:25:22.762165564    1284 fork_posix.cc:70]           Fork support is only compatible with the epoll1 and poll polling strategies


## Expenditure information from the Enterprise Data Link data Linda pulled.

In [2]:
enterprise = "Enterprise.xlsx"
enterprise_df = pd.read_excel(f"{GCS_FILE_PATH}{enterprise}")
enterprise_df = to_snakecase(enterprise_df)

In [3]:
enterprise_df = enterprise_df.rename(columns = {'project':'project_id'})

In [4]:
#keep only columns necessary.
enterprise_df = enterprise_df[['fy','cycle','project_id','project_name','tot_exp']]

In [5]:
enterprise_df.head()

fy cycle  project_id                                project_name  \
0  2016   One    16000007  Refurbishment of Seven Light Rail Vehicles   
1  2016   One    16000008   South Bay Bus Rapid Transit (BRT) Project   
2  2016   One    16000041                Bravo! Route 560 Rapid Buses   
3  2017   One    16000007  Refurbishment of Seven Light Rail Vehicles   
4  2017   One    16000008   South Bay Bus Rapid Transit (BRT) Project   

      tot_exp  
0  2182302.44  
1    30770.46  
2  1972865.52  
3  3691437.55  
4  1646527.16

In [6]:
enterprise_df.shape

(254, 5)

In [7]:
f'There are {enterprise_df.project_id.nunique()} unique project IDS'

'There are 106 unique project IDS'

In [8]:
enterprise_df.isna().sum()

fy              0
cycle           0
project_id      0
project_name    0
tot_exp         0
dtype: int64

#### Aggregate by project name and each cycle],so each project has only one record for total expenses

####  <font color='red'> 88 projects but only 74 in TIRCP</font> 

In [9]:

enterprise_df_group = enterprise_df.groupby(['project_name','cycle']).agg({'tot_exp':'sum'}).reset_index()

In [10]:
enterprise_df_group.project_name.nunique()

88

In [11]:
enterprise_df_group.shape

(88, 3)

In [12]:
enterprise_df_group

project_name  cycle  \
0                       #Electrify Anaheim (ZEV Bus Procurement)  Three   
1    #Electrify Anaheim: Changing the Transit Paradigm in So Cal  Three   
2     #Electrify Anaheim: Changing the Transit Paradigm in So.CA  Three   
3   (2018: 11) All Aboard 2018: Transform SoCal Rail T (Fencing)  Three   
4                     ACE Near-Term Capacity Improvement Program    Two   
5                                      ACE Wayside Power Project    One   
6   AVTA Component- Southern California Zero Emission Regional C    Two   
7    Accelerating Rail Modernization and Expansion (LRV Procure)  Three   
8                All Aboard 2018 (On-Time Performance Incentive)  Three   
9                         All Aboard 2018 (Right of Way Fencing)  Three   
10     All Aboard: Transforming SoCal Rail (Network Integration)    Two   
11               All Aboard: Transforming SoCal Rail (Timetable)    Two   
12                       Altamont Corridor Express Wayside Power    One   
13                 Blue Line Rail Corridor Enhancements (Design)  Three   
14    Blue Line Rail Corridor Enhancements (Network Integration)  Three   
15   Blue Line Rail Corridor Transit Enhancements (Rte 925 Infr)  Three   
16                                  Bravo! Route 560 Rapid Buses    One   
17  CCJPA Northern Corridor Enhancement Program (CALITP Phase 2)  Three   
18                Capitol Corridor Travel Time Reduction Project    One   
19                                   Carlsbad Poinsettia Station    Two   
20  Coastal Express/Pacific Surfliner Peak Hour Service Expansio  Three   
21              DMU to ZEMU Vehicle Conversion (DMU Procurement)  Three   
22                        DMU to ZEMU Vehicle Conversion Project  Three   
23   Dublin/Pleasanton Capac. Imprvmt. Congestion Reduct. Prgrm.  Three   
24                                 Elvira to Morena Double Track    Two   
25                        Elvira to Morena Double Track (R362GB)    Two   
26               From the Desert to the Sea (AVTA Bus Component)  Three   
27              From the Desert to the Sea (Network Integration)  Three   
28                                            Goleta Train Depot  Three   
29                     Goleta Train Depot (Eletric Shuttles-CON)  Three   
30                                    Goleta Train Depot (PA&ED)  Three   
31  Increased Rail Service to Roseville and Service optimization    Two   
32     Increased Rail Service to Roseville, Service Optimization    Two   
33                           Inglewood Transit Connector Project   Four   
34           LA Metro Blue Line Operational Improvements Project    One   
35                Light Rail Modernization and Expansion Program    Two   
36     Metro Red Line and Purple Line Core Capacity Improvements    Two   
37      Metropolitan Rapid Transit and Rail Connectivity Project    Two   
38              Monterey Bay Operations and Maintenance Facility    One   
39                NCTD Signal Op Signal Train Ctrl Repl Legacy 1  Three   
40                                           Network Integration  Three   
41              OC Street Car & OCTA System-wide Mobile Tracking    Two   
42            OC Streetcar and OCTA System-Wide Mobile Ticketing    Two   
43                         PS&E Soltrans Electric Infrastructure  Three   
44                    Pacific Surfliner Transit Transfer Program    One   
45          Peninsula Corridor Elecrtification Expansion Project  Three   
46          Peninsula Corridor Electrification Expansion Project  Three   
47                    Peninsula Corridor Electrification Project    Two   
48    Penisula Corridor Electrification Expansion Project (PLAT)  Three   
49  Procure Nine Fuel-Efficient, Tier VI EMD F-125 Locomotives f    One   
50     Purchase Zero Emission Buses to Support Transbay Tomorrow  Three   
51                             Rail Extension to Monterey County  Three   
52                               Redlands Passenger Rail Project    Two   
53

#### Aggregate by project ID and each cycle, so each project has only one record for total expenses

In [13]:
enterprise_df_project_id = enterprise_df.groupby(['project_id','cycle']).agg({'tot_exp':'sum'}).reset_index()

In [14]:
enterprise_df_project_id.shape

(106, 3)

In [15]:
enterprise_df_project_id

project_id  cycle      tot_exp
0      16000007    One   6315208.89
1      16000008    One   4000000.00
2      16000009    One  38023039.68
3      16000041    One   2320000.00
4      16000048    One  22284205.53
5      16000119    One    277840.47
6      16000121    One  41181000.00
7      16000188    One    713000.00
8      16000237    One  11000000.00
9      16000238    One   6841000.00
10     16000275    One  10000000.00
11     16000276    One   4619999.90
12     16000277    One     15000.00
13     16000329    One  38494000.00
14     17000040    One  27463000.00
15     17000044    Two   2413997.30
16     17000077    Two   4083000.00
17     17000128    Two   6481000.00
18     17000129    Two   2449000.00
19     17000130    Two  19727081.56
20     17000174    One   3760000.00
21     17000182    Two  49995000.00
22     17000183    Two  16100400.95
23     17000205    Two  45092000.00
24     17000206    Two    315332.37
25     17000234    Two   4017000.00
26     17000240    Two   3101000.00
27     18000170    One    185000.00
28     18000175    Two  11388000.00
29     18000237    Two  12705054.75
30     18000287    Two    428492.45
31     18000321  Three   1742773.01
32     18000323  Three    399996.08
33     18000324  Three  10025763.96
34     18000354  Three   6000000.00
35     18000355  Three   7835275.53
36     18000356  Three   2926950.80
37     18000357  Three  26867000.00
38     18000358  Three  19107952.73
39     18000359    Two    120000.00
40     19000014  Three    170791.14
41     19000017  Three    610535.78
42     19000021  Three    922975.95
43     19000063  Three   6656313.82
44     19000064  Three     71539.79
45     19000068    Two    421377.91
46     19000069    Two    421378.17
47     19000073  Three   6921214.22
48     19000075  Three     58994.67
49     19000078  Three    500000.00
50     19000079  Three  57306844.42
51     19000081  Three   5875000.00
52     19000086    Two   9204000.00
53     19000087  Three   2002556.00
54     19000088  Three      1219.13
55     19000089  Three    766093.67
56     19000090  Three    590000.00
57     19000091  Three    250000.00
58     19000092  Three    300000.00
59     19000093  Three    486509.70
60     19000111  Three    343488.02
61     19000120  Three  15950328.48
62     19000121  Three   2810949.37
63     19000122  Three   7508811.24
64     19000123  Three   5051375.56
65     19000128  Three  13492703.78
66     19000134  Three    382718.05
67     19000135  Three    522030.21
68     19000237  Three    217692.75
69     19000238  Three    178702.70
70     19000239  Three    215500.00
71     19000341  Three    237234.04
72     19000460  Three  15500000.00
73     20000003  Three         0.00
74     20000004  Three    731222.84
75     20000007  Three  31090378.54
76     20000042  Three   1726121.55
77     20000083  Three  12505114.06
78     20000084  Three    159628.42
79     20000133  Three     12562.03
80     20000156  Three    372000.00
81     20000160  Three   1153502.74
82     20000161  Three   1037021.58
83     20000190  Three   8757820.93
84     20000212  Three   3800000.00
85     20000214  Three     44463.76
86     20000218  Three     63708.00
87     20000275   Four  13321538.34
88     20000279   Four       981.93
89     20000280  Three    963246.92
90     20000287  Three     60448.38
91     21000080   Four    191083.15
92     21000084   Four    644919.97
93     21000087   Four    398867.26
94     21000145  Three       572.07
95     21000148  Three     11506.73
96     21000149  Three      4842.10
97     21000157  Three     63882.78
98     21000158  Three     10553.85
99     21000160  Three     73822.87
100    21000202  Three     16385.26
101    21000203  Three     12451.92
102    21000210  Three    127154.14
103    21000211  Three    635000.00
104    21000212  Three    240000.00
105    21000274  Three    271341.11

## Merge in with allocation sheet of Excel workbook 
* There are 171 unique project IDS in our allocation sheet compared to 106 project IDS in Data Link....
* Lots of missing Project IDS: 107 out of 307 rows of data.

In [16]:
#read in allocation df because that's the sheet with project ids. 
allocation = TIRCP_functions.allocation()

In [17]:
allocation.shape

(307, 49)

In [18]:
allocation.isna().sum()

Award_Year                                  0
Project_#                                   1
Award_Recipient                             1
Implementing_Agency                         1
Project_ID                                107
EA                                        122
Components                                  1
Phase                                       2
Allocation_Amt_Allocation_Sheet             0
Expended_Amt_Allocation_Sheet               0
SB1_Funding                                 0
SB1_Budget_Year                           140
GGRF_Funding                                0
GGRF_Budget_Year                          307
CTC_Financial_Resolution                  100
CTC_Allocation_Amendment                  307
CTC_Waiver                                307
CTC_CalSTA_Waiver                         307
PSA_#                                     113
CT_Document_#                             177
Date_Branch_Chief_Receives_PSA            304
Date_Regional_Coordinator_Receives

In [19]:
#keep only columns that are relevant.
allocation_df = allocation[['Award_Year','Project_ID','PPNO']]

In [20]:
f'There are {allocation_df.Project_ID.nunique()} unique project ids in allocation'

'There are 171 unique project ids in allocation'

In [21]:
#only keep rows with Project IDS
allocation_df = allocation_df.loc[allocation_df['Project_ID'].notnull()]

In [22]:
allocation_df.shape

(200, 3)

In [23]:
joined_alloc = pd.merge(allocation_df, enterprise_df_project_id, 
                        left_on='Project_ID', right_on='project_id', 
                        how = 'outer', indicator = True)

In [24]:
joined_alloc._merge.value_counts()

both          123
left_only      77
right_only      2
Name: _merge, dtype: int64

In [25]:
joined_alloc.shape

(202, 7)

In [26]:
joined_alloc.head(100)

Award_Year  Project_ID   PPNO  project_id  cycle      tot_exp     _merge
0       2015.0    16000048  CP005  16000048.0    One  22284205.53       both
1       2015.0    16000276  CP012  16000276.0    One   4619999.90       both
2       2016.0    16000276  CP023  16000276.0    One   4619999.90       both
3       2015.0    16000329  CP015  16000329.0    One  38494000.00       both
4       2015.0    16000119  CP007  16000119.0    One    277840.47       both
5       2015.0    16000275  CP013  16000275.0    One  10000000.00       both
6       2015.0    16000041  CP004  16000041.0    One   2320000.00       both
7       2015.0    16000007  CP001  16000007.0    One   6315208.89       both
8       2015.0    16000008  CP003  16000008.0    One   4000000.00       both
9       2015.0    16000188  CP008  16000188.0    One    713000.00       both
10      2015.0    17000174  CP008  17000174.0    One   3760000.00       both
11      2015.0    17000040  CP008  17000040.0    One  27463000.00       both
12      2015.0    16000121  CP006  16000121.0    One  41181000.00       both
13      2015.0    16000277  CP014  16000277.0    One     15000.00       both
14      2015.0    18000170  CP014  18000170.0    One    185000.00       both
15      2015.0    16000238  CP011  16000238.0    One   6841000.00       both
16      2015.0    16000009  CP002  16000009.0    One  38023039.68       both
17      2015.0    16000237  CP010  16000237.0    One  11000000.00       both
18      2016.0    17000128  CP018  17000128.0    Two   6481000.00       both
19      2016.0    17000129  CP019  17000129.0    Two   2449000.00       both
20      2016.0    17000181  CP023         NaN    NaN          NaN  left_only
21      2016.0    17000206  CP023  17000206.0    Two    315332.37       both
22      2016.0    17000206  CP023  17000206.0    Two    315332.37       both
23      2016.0    16000233  CP023         NaN    NaN          NaN  left_only
24      2016.0    16000233  CP023         NaN    NaN          NaN  left_only
25      2016.0    16000233  CP023         NaN    NaN          NaN  left_only
26      2016.0    16000233  CP023         NaN    NaN          NaN  left_only
27      2016.0    16000233  CP023         NaN    NaN          NaN  left_only
28      2016.0    18000359  CP023  18000359.0    Two    120000.00       both
29      2016.0    18000359  CP023  18000359.0    Two    120000.00       both
30      2016.0         TBD  CP023         NaN    NaN          NaN  left_only
31      2018.0         TBD  CP036         NaN    NaN          NaN  left_only
32      2018.0         TBD  CP036         NaN    NaN          NaN  left_only
33      2020.0         TBD  CP061         NaN    NaN          NaN  left_only
34      2016.0    21000266  CP076         NaN    NaN          NaN  left_only
35      2016.0    17000077  CP016  17000077.0    Two   4083000.00       both
36      2016.0   200000215  CP016         NaN    NaN          NaN  left_only
37      2016.0    20000197  CP058         NaN    NaN          NaN  left_only
38      2016.0    17000130  CP020  17000130.0    Two  19727081.56       both
39      2016.0    17000240  CP020  17000240.0    Two   3101000.00       both
40      2016.0    20000081  CP020         NaN    NaN          NaN  left_only
41      2016.0    19000068  CP043  19000068.0    Two    421377.91       both
42      2016.0    19000069  CP043  19000069.0    Two    421378.17       both
43      2016.0    21000152  CP043         NaN    NaN          NaN  left_only
44      2016.0    21000154  CP043         NaN    NaN          NaN  left_only
45      2016.0    17000182  CP021  17000182.0    Two  49995000.00       both
46      2016.0    18000175  CP021  18000175.0    Two  11388000.00       both
47      2016.0    17000234  CP024  17000234.0    Two   4017000.00       both
48      2016.0    20000211  CP024         NaN    NaN          NaN  left_only
49      2016.0    17000044  CP017  17000044.0    Two   2413997.30       both
50      2016.0    18000237  CP017  18000237.0    Two  12705054.75       both


In [27]:
#only keep "both" results...
joined_alloc2 = joined_alloc.loc[joined_alloc['_merge'] == 'both'].drop(columns =['_merge'])

In [28]:
#group by so that only one row for each PPNO number & Year
joined_alloc3 = (joined_alloc2
                  .groupby(['PPNO', 'Award_Year'])
                  .agg({'tot_exp':'sum'})
                  .reset_index()
                 )

In [29]:
joined_alloc3.shape

(51, 3)

In [30]:
#rename cols before joining with "joined" df 
joined_alloc3 = joined_alloc3.add_prefix('from_joined_alloc_df_')

In [31]:
joined_alloc3

from_joined_alloc_df_PPNO  from_joined_alloc_df_Award_Year  \
0                       1155                           2018.0   
1                       1230                           2016.0   
2                      2320B                           2018.0   
3                      CP001                           2015.0   
4                      CP002                           2015.0   
5                      CP003                           2015.0   
6                      CP004                           2015.0   
7                      CP005                           2015.0   
8                      CP006                           2015.0   
9                      CP006                           2016.0   
10                     CP006                           2018.0   
11                     CP007                           2015.0   
12                     CP008                           2015.0   
13                     CP010                           2015.0   
14                     CP011                           2015.0   
15                     CP012                           2015.0   
16                     CP013                           2015.0   
17                     CP014                           2015.0   
18                     CP015                           2015.0   
19                     CP016                           2016.0   
20                     CP017                           2016.0   
21                     CP018                           2016.0   
22                     CP019                           2016.0   
23                     CP020                           2016.0   
24                     CP021                           2016.0   
25                     CP022                           2016.0   
26                     CP023                           2016.0   
27                     CP024                           2016.0   
28                     CP025                           2016.0   
29                     CP027                           2018.0   
30                     CP028                           2018.0   
31                     CP031                           2018.0   
32                     CP032                           2018.0   
33                     CP033                           2018.0   
34                     CP034                           2018.0   
35                     CP035                           2018.0   
36                     CP036                           2018.0   
37                     CP039                           2018.0   
38                     CP041                           2018.0   
39                     CP043                           2016.0   
40                     CP046                           2018.0   
41                     CP047                           2018.0   
42                     CP048                           2018.0   
43                     CP051                           2018.0   
44                     CP053                           2018.0   
45                     CP055                           2018.0   
46                     CP061                           2020.0   
47                     CP062                           2020.0   
48                     CP069                           2020.0   
49                     CP072                           2020.0   
50                     CP079                           2018.0   

    from_joined_alloc_df_tot_exp  
0                      486509.70  
1                     9204000.00  
2                      500000.00  
3                     6315208.89  
4                    38023039.68  
5                     4000000.00  
6                     2320000.00  
7                    22284205.53  
8                    41181000.00  
9                    45092000.00  
10                   26867000.00  
11                     277840.47  
12                   31936000.00  
13                   11000000.00  
14                    6841000.00  
15                    4619999.90  
16                   100000

In [32]:
joined_alloc3.astype({'from_joined_alloc_df_Award_Year': 'int64'}).dtypes

from_joined_alloc_df_PPNO           object
from_joined_alloc_df_Award_Year      int64
from_joined_alloc_df_tot_exp       float64
dtype: object

## Joining original sheets on Award Year & PPNO

In [33]:
project = TIRCP_functions.project()

In [34]:
project.columns

Index(['Award_Year', 'Project_#', 'Local_Agency', 'Vendor_ID_#',
       'Project_Title', 'District', 'County', 'Key_Project_Elements',
       'Master_Agreement_Number', 'Master_Agreement_Expiration_Date',
       'Project_Manager', 'Regional_Coordinator',
       'Technical_Assistance-CALTP_(Y/N)', 'Technical_Assistance-Fleet_(Y/N)',
       'Technical_Assistance-Network_Integration_(Y/N)',
       'Technical_Assistance-Priority_Population_(Y/N)', 'Total_Project_Cost',
       'TIRCP_project_sheet', 'Allocated_Amount',
       'Unallocated_amt_project_sheet', 'Percentge_Allocated',
       'Expended_Amt_project_sheet', 'Other_Funds_Involved', 'Award_Cycle',
       'Local_Agency_Address', 'Local_Agency_City', 'Local_Agency_Zip',
       'Local_Agency_Contact', 'Local_Agency_Email',
       'Local_Agency_Phone_Number', 'Comments/Additional_Contacts', 'PPNO'],
      dtype='object')

In [35]:
project.dtypes

Award_Year                                                 int64
Project_#                                                  int64
Local_Agency                                              object
Vendor_ID_#                                               object
Project_Title                                             object
District                                                  object
County                                                    object
Key_Project_Elements                                      object
Master_Agreement_Number                                   object
Master_Agreement_Expiration_Date                  datetime64[ns]
Project_Manager                                           object
Regional_Coordinator                                      object
Technical_Assistance-CALTP_(Y/N)                          object
Technical_Assistance-Fleet_(Y/N)                          object
Technical_Assistance-Network_Integration_(Y/N)            object
Technical_Assistance-Prio

In [36]:
final_join = pd.merge(project, joined_alloc3, 
                      left_on = ['PPNO', 'Award_Year'], 
                      right_on = ['from_joined_alloc_df_PPNO', 'from_joined_alloc_df_Award_Year'],
                      how = 'left', indicator = True)

In [37]:
final_join._merge.value_counts()

both          49
left_only     25
right_only     0
Name: _merge, dtype: int64

#### Check to see if PPNO matches  cross original project sheet & joined 

In [38]:
final_join.shape

(74, 36)

In [39]:
final_join.PPNO.nunique()

70

In [40]:
project.PPNO.nunique()

70

In [41]:
PPNO_project = set(project.PPNO.unique().tolist())
PPNO_join = set(final_join.PPNO.unique().tolist())
PPNO_join - PPNO_project 

set()

In [42]:
final_join.columns

Index(['Award_Year', 'Project_#', 'Local_Agency', 'Vendor_ID_#',
       'Project_Title', 'District', 'County', 'Key_Project_Elements',
       'Master_Agreement_Number', 'Master_Agreement_Expiration_Date',
       'Project_Manager', 'Regional_Coordinator',
       'Technical_Assistance-CALTP_(Y/N)', 'Technical_Assistance-Fleet_(Y/N)',
       'Technical_Assistance-Network_Integration_(Y/N)',
       'Technical_Assistance-Priority_Population_(Y/N)', 'Total_Project_Cost',
       'TIRCP_project_sheet', 'Allocated_Amount',
       'Unallocated_amt_project_sheet', 'Percentge_Allocated',
       'Expended_Amt_project_sheet', 'Other_Funds_Involved', 'Award_Cycle',
       'Local_Agency_Address', 'Local_Agency_City', 'Local_Agency_Zip',
       'Local_Agency_Contact', 'Local_Agency_Email',
       'Local_Agency_Phone_Number', 'Comments/Additional_Contacts', 'PPNO',
       'from_joined_alloc_df_PPNO', 'from_joined_alloc_df_Award_Year',
       'from_joined_alloc_df_tot_exp', '_merge'],
      dtype='object

### Checking out our 2 columns of expended amounts:
* From original project data sheet
* Second merge of Data Link with Allocations df

In [43]:
final_join[['Award_Year','PPNO','Project_Title', 'TIRCP_project_sheet',
            'Expended_Amt_project_sheet', 'from_joined_alloc_df_tot_exp', '_merge']]

Award_Year   PPNO  \
0         2015  CP005   
1         2015  CP012   
2         2015  CP015   
3         2015  CP007   
4         2015  CP013   
5         2015  CP004   
6         2015  CP001   
7         2015  CP003   
8         2015  CP008   
9         2015  CP006   
10        2015  CP014   
11        2015  CP011   
12        2015  CP002   
13        2015  CP010   
14        2016  CP019   
15        2016  CP023   
16        2016  CP076   
17        2016  CP016   
18        2016  CP058   
19        2016  CP020   
20        2016  CP021   
21        2016  CP017   
22        2016  CP022   
23        2016  CP080   
24        2016   1230   
25        2016  CP006   
26        2016  CP025   
27        2016  CP057   
28        2016  CP026   
29        2018  2320B   
30        2018  CP027   
31        2018  CP028   
32        2018  CP055   
33        2018  CP036   
34        2018  CP079   
35        2018  CP029   
36        2018  CP054   
37        2018  CP051   
38        2018  CP030   
39        2018  CP031   
40        2018  CP031   
41        2018  CP039   
42        2018  CP053   
43        2018  CP034   
44        2018  CP077   
45        2018  CP032   
46        2018  CP006   
47        2018  CP035   
48        2018  CP078   
49        2018  CP048   
50        2018  CP047   
51        2018  CP057   
52        2018  CP045   
53        2018  CP046   
54        2018  CP041   
55        2018  CP033   
56        2018   1155   
57        2020  CP059   
58        2020  CP060   
59        2020  CP061   
60        2020  CP062   
61        2020  CP063   
62        2020  CP064   
63        2020  CP065   
64        2020  CP066   
65        2020  CP067   
66        2020  CP068   
67        2020  CP069   
68        2020  CP070   
69        2020  CP071   
70        2020  CP072   
71        2020  CP073   
72        2020  CP074   
73        2020  CP075   

                                                                                                           Project_Title  \
0                                                         Regional Transit Interconnectivity & Environmental Sustability   
1                                                                                          Travel Time Reduction Project   
2                                 Willowbrook/Rosa Parks Station & Blue Line Light Rail Operational Improvements Project   
3                                                                             Pacific Surfliner Transit Transfer Program   
4                                       Monterey Bay Operations and Maintenance Facility/Salinas Transit Service Project   
5                                                                                           Bravo! Route 560 Rapid Buses   
6                                                   Sacramento Regional Transit's Refurbishment of 7 Light Rail Vehicles   
7                                                                                            South Bay Bus Rapid Transit   
8                                                    San Diego Metropolitan Transit System Trolley Capacity Improvements   
9                                                                               SFMTA Light Rail Vehicle Fleet Expansion   
10                                                                                             ACE Wayside Power Project   
11                                          Bus Rapid Transit – Martin Luther King Corridor and Crosstown Miner Corridor   
12                                                         Purchase of Nine Fuel-Efficient Tier IV EMD F-125 Locomotives   
13                                                                                       SMART Rail Car Capacity Project   
14                  Zero Emission Bus and Vanpool Expansion in the Antelope Valley, Kern County and the Coachella Valley   
15                               Increased Rail Service to Roseville, Service Optimization and Standby Power Investments   
16                  

### Still missing 25 values - Manually replace them.
* Manual replacement.

In [44]:
#subset final joined data frame for projects that still have N/A in expenditures 
final_join2_subset = final_join[['Award_Year','PPNO','Project_Title', 'from_joined_alloc_df_tot_exp']]
Nulls = final_join2_subset.loc[final_join2_subset['from_joined_alloc_df_tot_exp'].isnull()]

In [45]:
#Read in crosswalk
crosswalk_expenditures = pd.read_excel('Unmatched_Data_Crosswalk.xlsx')

In [46]:
#Merge in Crosswalk 
final_join2 = pd.merge(final_join, crosswalk_expenditures, on = ["Award_Year", "PPNO","Project_Title"], how = "left")

In [47]:
#fill in NaN in original expenses column with dummy variable 888
final_join2['from_joined_alloc_df_tot_exp'] = final_join2['from_joined_alloc_df_tot_exp'].fillna(888)

In [48]:
#Map Crosswalk 
final_join2.from_joined_alloc_df_tot_exp = (final_join2.apply(lambda x: x.Expenses_Manual 
                                                              if x.from_joined_alloc_df_tot_exp == 888.0 
                                                              else x.from_joined_alloc_df_tot_exp, axis=1)
                                           )

### Some expended values aren't really correct compared to SAR

In [49]:
final_join2.loc[(final_join2['PPNO'] == 'CP023'), "from_joined_alloc_df_tot_exp"] =   8999000 
final_join2.loc[(final_join2['PPNO'] == 'CP036'), "from_joined_alloc_df_tot_exp"] =  80340000 
#CP030 Cycle 3 2018 Lossan
final_join2.loc[(final_join2['PPNO'] == 'CP030'), "from_joined_alloc_df_tot_exp"] =    1088499000 

## Fill in $0 TIRCP Amount
* CP045
* CP046
* CP078

In [50]:
final_join2.loc[(final_join2['PPNO'] == 'CP045'), "TIRCP_project_sheet"] =  8641000 
final_join2.loc[(final_join2['PPNO'] == 'CP046'), "TIRCP_project_sheet"] = 10788000 
final_join2.loc[(final_join2['PPNO'] == 'CP078'), "TIRCP_project_sheet"] =  15000000 


## Missing allocated amounts & unallocated - fill in w/ TIRCP amount.
* Clarified with Linda that these should be filled in with TIRCP amount

In [51]:
missing_allocation = final_join2.loc[final_join2['Allocated_Amount'] == 0] 

In [52]:
missing_allocation = missing_allocation[['Project_Title','PPNO','TIRCP_project_sheet', 
                    'Unallocated_amt_project_sheet', 
                    'Allocated_Amount','from_joined_alloc_df_tot_exp']]

In [53]:
#export missing allocated amounts to csv to crosswalk in. 
#missing_allocation = missing_allocation.to_excel('missing_allocated_amounts.xlsx')

In [54]:
#Read in crosswalk
crosswalk_allocated = pd.read_excel('Unmatched_Allocated_Crosswalk.xlsx')

In [55]:
crosswalk_allocated = crosswalk_allocated.drop(columns = ['Unallocated_amt_manual',
                                                          'from_joined_alloc_df_tot_exp']) 

In [56]:
#Merge in Crosswalk 
final_join2 = pd.merge(final_join2, crosswalk_allocated, on = ["PPNO","Project_Title"], how = "left")

In [57]:
#Map Crosswalk 
final_join2.Allocated_Amount = (final_join2.apply(lambda x: x.Allocated_Amount_Manual 
                                                              if x.Allocated_Amount == 0 
                                                              else x.Allocated_Amount, axis=1)
                                           )

In [58]:
#Check that everything mapped over: allocated, expenses, TIRCP
final_join2[['PPNO','TIRCP_project_sheet',
             'Allocated_Amount', 'from_joined_alloc_df_tot_exp']].sort_values('PPNO')

PPNO  TIRCP_project_sheet  Allocated_Amount  from_joined_alloc_df_tot_exp
56   1155         1.014800e+07          500000.0                  4.865097e+05
24   1230         9.204000e+06         9204000.0                  9.204000e+06
29  2320B         1.400000e+07         8000000.0                  5.000000e+05
6   CP001         6.427000e+06         6427000.0                  6.315209e+06
12  CP002         4.118100e+07        41181000.0                  3.802304e+07
7   CP003         4.000000e+06         4000000.0                  4.000000e+06
5   CP004         2.320000e+06         2320000.0                  2.320000e+06
0   CP005         2.440300e+07        24403000.0                  2.228421e+07
25  CP006         4.509200e+07        45092000.0                  4.509200e+07
46  CP006         2.686700e+07        26867000.0                  2.686700e+07
9   CP006         4.118100e+07        41181000.0                  4.118100e+07
3   CP007         1.675000e+06         1675000.0                  2.778405e+05
8   CP008         3.198600e+07        31936000.0                  3.193600e+07
13  CP010         1.100000e+07        11000000.0                  1.100000e+07
11  CP011         6.841000e+06         6841000.0                  6.841000e+06
1   CP012         4.620000e+06         4620000.0                  4.620000e+06
4   CP013         1.000000e+07        10000000.0                  1.000000e+07
10  CP014         2.000000e+05          200000.0                  2.000000e+05
2   CP015         3.849400e+07        38494000.0                  3.849400e+07
17  CP016         8.000000e+06         8000000.0                  4.083000e+06
21  CP017         2.800000e+07        28000000.0                  1.511905e+07
14  CP019         6.481000e+06         8930000.0                  2.449000e+06
19  CP020         6.920900e+07        69446000.0                  2.282808e+07
20  CP021         8.200000e+07        74460000.0                  6.138300e+07
22  CP022         2.000000e+07        20000000.0                  1.610040e+07
15  CP023         8.999000e+06         3446000.0                  8.999000e+06
26  CP025         1.645900e+07         8000000.0                  4.284925e+05
28  CP026         4.000000e+08       102899816.0                  1.675077e+07
30  CP027         2.861700e+07        28617000.0                  1.582936e+07
31  CP028         9.205000e+06         9205000.0                  6.727854e+06
35  CP029         3.610400e+07        36104000.0                  7.798000e+06
38  CP030         1.088499e+09       261200000.0                  1.088499e+09
40  CP031         9.136600e+07        60820000.0                  2.083520e+07
39  CP031         2.296200e+07        25562000.0                  2.083520e+07
45  CP032         4.009800e+07        32882000.0                  1.643395e+06
55  CP033         8.757080e+08       364207000.0                  3.349930e+07
43  CP034         3.000000e+07        28500000.0                  2.219430e+07
47  CP035         5.005000e+08        65820000.0                  2.034686e+07
33  CP036         8.034000e+07        50040000.0                  8.034000e+07
41  CP039         2.036380e+08       125682000.0                  6.276571e+07
54  CP041         2.100000e+07        21000000.0                  1.927874e+07
52  CP045         8.641000e+06         8641000.0                  8.641000e+06
53  CP046         1.078800e+07         9225000.0                  7.244003e+06
50  CP047         1.300900e+07         2200000.0                  1.571223e+06
49  CP048         9.600000e+06         9600000.0                  4.144464e+06
37  CP051         2.050000e+07        20500000.0                  3.827180e+05
42  CP053         6.435000e+07        53182000.0                  3.132761e+07
36  CP054         3.050000e+06         3050000.0                  3.050000e+06
32  CP055         3.186000e+08       368600000.0                  3.751534e+07
51  CP057         7.300000e+08       730000000.0       

## Running new dataframe with my Tableau function.

In [59]:
#dropping cols I don't need
final_join3 = final_join2.drop(columns=['Expended_Amt_project_sheet',
                                       'from_joined_alloc_df_PPNO',
                                        'from_joined_alloc_df_Award_Year', '_merge', 'Unallocated_amt_project_sheet',
                                       'Allocated_Amount_Manual'])

In [60]:
#just renaming the Data Link expenditures col to project sheet to make things easier 
final_join3 = final_join3.rename(columns = {'from_joined_alloc_df_tot_exp':'Expended_Amt_project_sheet'}) 

In [96]:
def tableau(df):
    #Keeping only the columns we want
    df = (df[['PPNO','Award_Year', 'Project_#', 'Local_Agency', 'Vendor_ID_#',
       'Project_Title', 'District', 'County', 'Key_Project_Elements',
       'Master_Agreement_Number', 'Master_Agreement_Expiration_Date',
       'Project_Manager', 'Regional_Coordinator',
       'Technical_Assistance-CALTP_(Y/N)', 'Technical_Assistance-Fleet_(Y/N)',
       'Technical_Assistance-Network_Integration_(Y/N)',
       'Technical_Assistance-Priority_Population_(Y/N)', 'Total_Project_Cost',
       'TIRCP_project_sheet', 'Allocated_Amount',
       'Percentge_Allocated',
       'Expended_Amt_project_sheet', 'Other_Funds_Involved']]
                 )
    #Getting percentages & filling in with 0
    df['Expended_Percent'] = df['Expended_Amt_project_sheet']/df['Allocated_Amount']
    df['Allocated_Percent'] = df['Allocated_Amount']/df['TIRCP_project_sheet']
    
    # Subtract TIRCP with Allocated Amount with Unallocated
    df['Unallocated_Amount'] = df["TIRCP_project_sheet"] - df["Allocated_Amount"]
    #filling in for 0's 
    df[['Expended_Percent','Allocated_Percent']] = df[['Expended_Percent','Allocated_Percent']].fillna(value=0)
   
    df[['Expended_Percent','Allocated_Percent']]  =  df[['Expended_Percent','Allocated_Percent']].replace(np.inf, 0)
    #Categorizing expended percentage into bins
    def expended_percent(row):
           
            if ((row.Expended_Percent > 0) and (row.Expended_Percent < .26)):
                return "1-25"
            elif ((row.Expended_Percent > .25) and (row.Expended_Percent < .51)):
                return "26-50"
            elif ((row.Expended_Percent > .50) and (row.Expended_Percent < .76)):
                return "51-75"
            elif ((row.Expended_Percent > .75) and (row.Expended_Percent < 1.0)):
                return "76-99"
            else:
                return "100"
    df["Expended_Percent_Group"] = df.apply(lambda x: expended_percent(x), axis=1)
    
    # Categorize years and expended_percent_group into bins
    def progress(df):   
        ### 2015 ### 
        if (df['Award_Year'] == 2015) and (df['Expended_Percent_Group'] == "1-25") | (df['Expended_Percent_Group'] == "26-50") | (df['Expended_Percent_Group'] == "51-75"):
            return 'Behind'
        elif (df['Award_Year'] == 2015) and (df['Expended_Percent_Group'] == "76-99"):
            return 'On Track'
        
        ### 2016 ###
        elif (df['Award_Year'] == 2016) and (df['Expended_Percent_Group'] == "1-25") | (df['Expended_Percent_Group'] == "26-50"):
            return 'Behind'
        elif (df['Award_Year'] == 2016) and (df['Expended_Percent_Group'] == "51-75") | (df['Expended_Percent_Group'] == "76-99"):
            return 'On Track'
        
        ### 2018 ###
        elif (df['Award_Year'] == 2018) and (df['Expended_Percent_Group'] == "1-25"):
            return 'Behind'
        elif (df['Award_Year'] == 2018) and (df['Expended_Percent_Group'] == "26-50") | (df['Expended_Percent_Group'] == "51-75"):
            return 'On Track'
        elif (df['Award_Year'] == 2018) and (df['Expended_Percent_Group'] == "76-99"):
            return 'Ahead'
        
        ### 2020 ### 
        elif (df['Award_Year'] == 2020) and (df['Expended_Percent_Group'] == "1-25"):
            return 'Behind'
        elif (df['Award_Year'] == 2020) and (df['Expended_Percent_Group'] == "26-50"):
            return 'On Track'
        elif (df['Award_Year'] == 2020) and (df['Expended_Percent_Group'] == "51-75") | (df['Expended_Percent_Group'] == "76-99"):
            return 'Ahead'

        ### Else ### 
        else: 
            return "100% of allocated funds spent"

    df['Progress'] = df.apply(progress, axis = 1)
    
    #Which projects are large,small, medium
    p75 = df.TIRCP_project_sheet.quantile(0.75).astype(float)
    p25 = df.TIRCP_project_sheet.quantile(0.25).astype(float)
    p50 = df.TIRCP_project_sheet.quantile(0.50).astype(float)
    
    def project_size (row):
        if ((row.TIRCP_project_sheet > 0) and (row.TIRCP_project_sheet < p25)):
            return "Small"
        elif ((row.TIRCP_project_sheet > p25) and (row.TIRCP_project_sheet < p75)):
            return "Medium"
        elif ((row.TIRCP_project_sheet > p50) and (row.TIRCP_project_sheet > p75 )):
            return "Large"
        else:
            return "$0 recorded for TIRCP"
        
    df["Project_Category"] = df.apply(lambda x: project_size(x), axis=1)
     #Rename cols to the right names
    df = (df.rename(columns = {'Expended_Amt_project_sheet':'Expended_Amount', 
                                                'TIRCP_project_sheet': "TIRCP_Amount"}
                  ))
    ### GCS ###
   # df = df.to_excel(f'{GCS_FILE_PATH}INTERIM_EXPENDITURE_Tableau_Parquet.xlsx')
  

    return df 

In [97]:
test = tableau(final_join3)

/tmp/ipykernel_1284/4081200191.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/tmp/ipykernel_1284/4081200191.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/tmp/ipykernel_1284/4081200191.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/tmp/ipykernel_1284/40

In [64]:
pd.options.display.float_format = '{:,}'.format
pd.options.display.float_format = "{:.2f}".format

In [98]:
test[['Award_Year','PPNO','Project_Title', 'TIRCP_Amount','Allocated_Amount','Unallocated_Amount','Expended_Amount','Expended_Percent','Progress']].sort_values('Expended_Percent')

Award_Year   PPNO  \
59        2020  CP061   
34        2018  CP079   
37        2018  CP051   
45        2018  CP032   
26        2016  CP025   
29        2018  2320B   
55        2018  CP033   
32        2018  CP055   
67        2020  CP069   
28        2016  CP026   
3         2015  CP007   
70        2020  CP072   
35        2018  CP029   
14        2016  CP019   
47        2018  CP035   
19        2016  CP020   
40        2018  CP031   
49        2018  CP048   
41        2018  CP039   
17        2016  CP016   
21        2016  CP017   
30        2018  CP027   
42        2018  CP053   
60        2020  CP062   
50        2018  CP047   
31        2018  CP028   
43        2018  CP034   
53        2018  CP046   
22        2016  CP022   
39        2018  CP031   
20        2016  CP021   
0         2015  CP005   
54        2018  CP041   
12        2015  CP002   
56        2018   1155   
6         2015  CP001   
1         2015  CP012   
58        2020  CP060   
71        2020  CP073   
51        2018  CP057   
69        2020  CP071   
66        2020  CP068   
65        2020  CP067   
52        2018  CP045   
64        2020  CP066   
63        2020  CP065   
62        2020  CP064   
68        2020  CP070   
57        2020  CP059   
36        2018  CP054   
46        2018  CP006   
2         2015  CP015   
4         2015  CP013   
5         2015  CP004   
7         2015  CP003   
8         2015  CP008   
9         2015  CP006   
10        2015  CP014   
11        2015  CP011   
13        2015  CP010   
48        2018  CP078   
73        2020  CP075   
18        2016  CP058   
23        2016  CP080   
24        2016   1230   
25        2016  CP006   
27        2016  CP057   
72        2020  CP074   
44        2018  CP077   
16        2016  CP076   
33        2018  CP036   
15        2016  CP023   
38        2018  CP030   
61        2020  CP063   

                                                                                                           Project_Title  \
59                                                                        Sacramento Valley Station (SVS) Transit Center   
34                                                                                  Southwest Fresno Community Connector   
37                                               Dublin/Pleasanton Capacity Improvement and Congestion Reduction Program   
45                                                                          Blue Line Rail Corridor Transit Enhancements   
26                                                                          ACE Near-Term Capacity Improvement Program\n   
29                      Purchase Zero Emission High Capacity Buses to Support Transbay Tomorrow and Clean Corridors Plan   
55                                                                  Southern California Optimized Rail Expansion (SCORE)   
32         The Transbay Corridor Core Capacity Program: Vehicle Acquistion and Communications-Based Train Control System   
67                                                                         SDConnect: San Diego Rail Improvement Program   
28                                                                          SB 132 ACE Extension Lathrop to Ceres/Merced   
3                                                                             Pacific Surfliner Transit Transfer Program   
70                                                                          Solano Regional Transit Improvements Phase 2   
35                            Los Angeles City: Leading the Transformation to Zero-Emission Electric Bus Transit Service   
14                  Zero Emission Bus and Vanpool Expansion in the Antelope Valley, Kern County and the Coachella Valley   
47                                                                                                           Valley Rail   
19                                                             Metro Red Line and Purple Line Core Capacity Improvements   
40                  

In [94]:
test.Progress.value_counts()

100% of allocated funds spent    37
On Track                         16
Behind                           15
Ahead                             6
Name: Progress, dtype: int64

In [100]:
test.TIRCP_Amount.describe()

count           74.00
mean      78363824.32
std      191325631.46
min         200000.00
25%        8160250.00
50%       18229500.00
75%       40910250.00
max     1088499000.00
Name: TIRCP_Amount, dtype: float64

In [99]:
test.to_excel(f'{GCS_FILE_PATH}tableau_with_temporary_expenditure_sol.xlsx', index = False)